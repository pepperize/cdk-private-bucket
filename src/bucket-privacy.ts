import { aws_s3, IAspect } from "aws-cdk-lib";
import { IConstruct } from "constructs";

export class BucketPrivacy implements IAspect {
  visit(node: IConstruct): void {
    if (node instanceof aws_s3.CfnBucket) {
      const bucket = node as aws_s3.CfnBucket;

      bucket.bucketEncryption = {
        serverSideEncryptionConfiguration: [
          {
            serverSideEncryptionByDefault: {
              kmsMasterKeyId: undefined,
              sseAlgorithm: "AES256",
            },
          },
        ],
      };

      bucket.publicAccessBlockConfiguration = {
        blockPublicAcls: true,
        blockPublicPolicy: true,
        ignorePublicAcls: true,
        restrictPublicBuckets: true,
      };
    }

    if (node instanceof aws_s3.CfnBucketPolicy) {
      const bucketPolicy = node as aws_s3.CfnBucketPolicy;
      console.log(bucketPolicy.policyDocument);
    }
  }
}
