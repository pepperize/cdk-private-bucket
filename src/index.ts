import { aws_s3 } from "aws-cdk-lib";
import { Construct } from "constructs";

export interface PrivateBucketProps extends aws_s3.BucketProps {}

export class PrivateBucket extends aws_s3.Bucket {
  constructor(scope: Construct, id: string, props: PrivateBucketProps = {}) {
    super(scope, id, {
      encryption: aws_s3.BucketEncryption.S3_MANAGED,
      ...props,
      blockPublicAccess: aws_s3.BlockPublicAccess.BLOCK_ALL,
      enforceSSL: true,
    });
  }
}
