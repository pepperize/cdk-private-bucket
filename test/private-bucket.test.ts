import { Capture, Template } from "@aws-cdk/assertions";
import { Stack } from "@aws-cdk/core";
import { PrivateBucket } from "../src";

describe("PrivateBucket", () => {
  it("should have encryption", () => {
    // Given
    const stack = new Stack();
    // When
    new PrivateBucket(stack, "id");
    // Then
    const template = Template.fromStack(stack);
    template.hasResourceProperties("AWS::S3::Bucket", {
      BucketEncryption: {
        ServerSideEncryptionConfiguration: [
          {
            ServerSideEncryptionByDefault: {
              SSEAlgorithm: "AES256",
            },
          },
        ],
      },
    });
  });
  it("should block public access", () => {
    // Given
    const stack = new Stack();
    // When
    new PrivateBucket(stack, "id");
    // Then
    const template = Template.fromStack(stack);
    template.hasResourceProperties("AWS::S3::Bucket", {
      PublicAccessBlockConfiguration: {
        BlockPublicAcls: true,
        BlockPublicPolicy: true,
        IgnorePublicAcls: true,
        RestrictPublicBuckets: true,
      },
    });
  });
  it("should enforce SSL", () => {
    // Given
    const stack = new Stack();
    // When
    new PrivateBucket(stack, "id");
    // Then
    const template = Template.fromStack(stack);
    const capture = new Capture();
    template.hasResourceProperties("AWS::S3::BucketPolicy", {
      PolicyDocument: {
        Statement: [capture],
      },
    });
    expect(capture.asObject()).toMatchObject({
      Action: "s3:*",
      Condition: {
        Bool: {
          "aws:SecureTransport": "false",
        },
      },
      Effect: "Deny",
      Principal: {
        AWS: "*",
      },
    });
  });
  it("should match snapshot", () => {
    // Given
    const stack = new Stack();
    // When
    new PrivateBucket(stack, "id");
    // Then
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
