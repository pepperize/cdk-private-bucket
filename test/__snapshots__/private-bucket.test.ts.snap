// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateBucket should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "id4D1900E5": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "idPolicyCB6F200B": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "id4D1900E5",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "id4D1900E5",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "id4D1900E5",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
